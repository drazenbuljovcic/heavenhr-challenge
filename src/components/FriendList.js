import React, { Component } from 'react';
import { PropTypes } from 'prop-types';

import FriendListItem from './FriendListItem';
import PaginationControls from './PaginationControls';

import styles from './FriendList.css';

class FriendList extends Component {
  //#region MARK: Super methods
  constructor(props, context) {
    super(props, context);

    this.handlePaginationToNextPage = this.handlePaginationToNextPage.bind(this);
    this.handlePaginationToPreviousPage = this.handlePaginationToPreviousPage.bind(this);
    this.handlePaginationToPageNumber = this.handlePaginationToPageNumber.bind(this);
    this.handlePaginationToFirstPage = this.handlePaginationToFirstPage.bind(this);
    this.handlePaginationToLastPage = this.handlePaginationToLastPage.bind(this);
  }

  componentWillMount() {
    this.props.refreshPaginationInfo(
      this.props.friendListInfo.friendsById,
      this.props.listItemCountPerPage,
    );
  }

  componentWillReceiveProps (nextProps) {
    const nextFriendsList = nextProps.friendListInfo.friendsById;
    const oldFriendsList = this.props.friendListInfo.friendsById;

    const { totalPages, currentPage } = nextProps.friendListInfo.paginationInfo;
    if(totalPages === currentPage && !!currentPage) {
      this.handlePaginationToPreviousPage(); return;
    }

    if (JSON.stringify(oldFriendsList) !== JSON.stringify(nextFriendsList)) {
      nextProps.refreshPaginationInfo(nextFriendsList, nextProps.listItemCountPerPage);
    }
  }
  //#endregion

  //#region MARK: Pagination navigation handlers4
  handlePaginationToFirstPage() {
    this.props.changePaginationPage(0);
  }
  
  handlePaginationToLastPage() {
    this.props.changePaginationPage(this.props.friendListInfo.paginationInfo.totalPages - 1);
  }

  handlePaginationToNextPage() {
    this.props.changePaginationPage(this.props.friendListInfo.paginationInfo.currentPage + 1);
  }

  handlePaginationToPreviousPage() {
    this.props.changePaginationPage(this.props.friendListInfo.paginationInfo.currentPage - 1);
  }

  handlePaginationToPageNumber(pageNumber) {
    this.props.changePaginationPage(pageNumber);
  }
  //#endregion

  render () {
    const { friendsById, paginationInfo } = this.props.friendListInfo;
    const { itemCount, currentPageFriendsList } = paginationInfo;

    const friendsList = friendsById.length <= itemCount ? friendsById : currentPageFriendsList;
    return (
      <div>
        <ul className={styles.friendList}>
          {
            friendsList.map((friend, index) => {
              return (
                <FriendListItem
                  key={index}
                  id={friend.id}
                  name={friend.name}
                  gender={friend.gender}
                  starred={friend.starred}
                  {...this.props.actions} />
              );
            })
          }
        </ul>
        { friendsById.length > itemCount && (
          <PaginationControls
            totalPages={paginationInfo.totalPages}
            currentPage={paginationInfo.currentPage}
            totalCount={friendsById.length}
            startingItemIndex={paginationInfo.startingItemIndex}
            endingItemIndex={paginationInfo.endingItemIndex}
            handlePaginationToFirstPage={this.handlePaginationToFirstPage}
            handlePaginationToLastPage={this.handlePaginationToLastPage}
            handlePaginationToNextPage={this.handlePaginationToNextPage}
            handlePaginationToPreviousPage={this.handlePaginationToPreviousPage}
            handlePaginationToPageNumber={this.handlePaginationToPageNumber}
          />
        )}
      </div>
    );
  }
}

FriendList.propTypes = {
  friendListInfo: PropTypes.object.isRequired,
  listItemCountPerPage: PropTypes.number.isRequired,
  refreshPaginationInfo: PropTypes.func.isRequired,
  changePaginationPage: PropTypes.func.isRequired,
  actions: PropTypes.object.isRequired
};

export default FriendList;
