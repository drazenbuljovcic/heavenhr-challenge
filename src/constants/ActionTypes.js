export const ADD_FRIEND = 'ADD_FRIEND';
export const STAR_FRIEND = 'STAR_FRIEND';
export const DELETE_FRIEND = 'DELETE_FRIEND';

export const REFRESH_PAGINATION = 'REFRESH_PAGINATION';
export const CHANGE_PAGINATION_PAGE = 'CHANGE_PAGINATION_PAGE';