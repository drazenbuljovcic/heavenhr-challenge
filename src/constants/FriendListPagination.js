export const ITEM_COUNT = 2;