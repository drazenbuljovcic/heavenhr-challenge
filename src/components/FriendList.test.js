import React from 'react';
import renderer from 'react-test-renderer';

import FriendList from './FriendList';
import { ITEM_COUNT } from '../constants/FriendListPagination';

describe('Friend List', () => {
  it('should match it\'s empty snapshot', () => {
    const component = renderer.create(
      <FriendList 
        listItemCountPerPage={ITEM_COUNT}
        refreshPaginationInfo={() => {}}
        changePaginationPage={() => {}}
        friendListInfo={{ friendsById: [], paginationInfo: { itemCount: 0 }}}
        actions={{}}
      />
    );
    expect(component).toMatchSnapshot();
  });
});