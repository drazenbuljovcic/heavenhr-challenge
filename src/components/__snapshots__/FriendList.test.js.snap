exports[`Friend List should match it's empty snapshot 1`] = `
<div>
  <ul
    className={undefined} />
</div>
`;
